dask/dataframe/partitionquantiles.py

```diff
@@ -0,0 +1,447 @@
+    return vals.tolist(), weights.tolist()
+
+
+def merge_and_compress_summaries(vals_and_weights):
+    """Merge and sort percentile summaries that are already sorted.
+
+    Each item is a tuple like ``(vals, weights)`` where vals and weights
+    are lists.  We sort both by vals.
+
+    Equal values will be combined, their weights summed together.
+    """
+    vals_and_weights = [x for x in vals_and_weights if x]
+    if not vals_and_weights:
+        return ()
+    it = merge_sorted(*[zip(x, y) for x, y in vals_and_weights])
+    vals = []
+    weights = []
+    vals_append = vals.append
+    weights_append = weights.append
+    val, weight = prev_val, prev_weight = next(it)
+    for val, weight in it:
+        if val == prev_val:
+            prev_weight += weight
+        else:
+            vals_append(prev_val)
+            weights_append(prev_weight)
+            prev_val, prev_weight = val, weight
+    if val == prev_val:
+        vals_append(prev_val)
+        weights_append(prev_weight)
+    return vals, weights
+
+
+def process_val_weights(vals_and_weights, npartitions, dtype_info):
+    """Calculate final approximate percentiles given weighted vals
+
+    ``vals_and_weights`` is assumed to be sorted.  We take a cumulative
+    sum of the weights, which makes them percentile-like (their scale is
+    [0, N] instead of [0, 100]).  Next we find the divisions to create
+    partitions of approximately equal size.
+
+    It is possible for adjacent values of the result to be the same.  Since
+    these determine the divisions of the new partitions, some partitions
+    may be empty.  This can happen if we under-sample the data, or if there
+    aren't enough unique values in the column.  Increasing ``upsample``
+    keyword argument in ``df.set_index`` may help.
+    """
+    vals, weights = vals_and_weights
+    vals = np.array(vals)
+    weights = np.array(weights)
+
+    # We want to create exactly `npartition` number of groups of `vals` that
+    # are approximately the same weight and non-empty if possible.  We use a
+    # simple approach (more accurate algorithms exist):
+    # 1. Remove all the values with weights larger than the relative
+    #    percentile width from consideration (these are `jumbo`s)
+    # 2. Calculate percentiles with "interpolation=left" of percentile-like
+    #    weights of the remaining values.  These are guaranteed to be unique.
+    # 3. Concatenate the values from (1) and (2), sort, and return.
+    #
+    # We assume that all values are unique, which happens in the previous
+    # step `merge_and_compress_summaries`.
+
+    if len(vals) <= npartitions + 1:
+        # The data is under-sampled.  Distribute the empty partitions.
+        duplicated_index = np.linspace(
+            0, len(vals) - 1, npartitions - len(vals) + 1, dtype=int
+        )
+        duplicated_vals = vals[duplicated_index]
+        rv = np.concatenate([vals, duplicated_vals])
+        rv.sort()
```

Comment:
If we have too few samples to work with, is there an easy way for us to create the `dask.dataframe` with fewer partitions instead of knowingly creating empty partitions?  Hmm, and I suppose we can use interpolation for number dtypes to estimate additional partition divisions.